WebUI: add multi-file renaming

loligans opened the pull request:
Closes #16239.
(Partially addresses #6961.)

The purpose of the PR is to enable users to bulk rename files on the WebUI. 

![Renaming 200 config files](https://user-images.githubusercontent.com/3218210/209455284-24cfc26d-fed0-426a-847a-914e6a15f286.gif)  
_Image Description: Renaming 200 config files using a named variable and the file enumeration variable_

## Search Options
**Search Field**: The text criteria to search for as either a plain text string or a regular expression.

**Use Regular Expressions**: When checked, the search field will be interpreted as an ECMAScript regular expression. If not then the search field will be interpreted as a plain text string.

**Match all occurrences**: When checked, all matches of text in the search field will be replaced with the replacement text. Otherwise, only the first instance will be replaced.

**Case Sensitive**: When checked, the text specified in the search field will be treated as case sensitive.

## Replace Options
**Replacement Field**: The replacement field is what you want to replace any matches with.

**Applies To**: 
 - **Filename + Extension**: The search string will be applied to the Filename and the Extension
 - **Filename**: The search string will only be applied to the Filename
 - **Extension**: The search string will only be applied to the Extension

**Include Files**: When checked, the search will apply to files regardless of any files that are checked

**Include Folders**: When checked, the search will apply to folders regardless of any folders are checked

**Enumerate Files**: Specifies the starting point for enumerating files. Files can be enumerated by specifying $d in the replacement field.

## Additional Capabilities
Right clicking a folder will allow you to toggle all items underneath that folder. `Check` vs. `Unchecked` is determined by the row you right clicked. i.e. if the row you right clicked is `unchecked` then it will be marked `checked` and all child items will be marked `checked` as well.

The Replacement Field has some special variables that can be used in the replacement field.
 - **$0**: The whole match
 - **$N**: Every capture group (Where **N** is an integer, i.e. **$1 $2 $3 $4** ... etc)
 - **$NamedGroup**: The value of named capture groups (Where **NamedGroup** is the actual name of your capture group)
   - Named groups can be specified by enabling RegEx and specifying a capture group i.e. `(?<mynamedgroup>myfilename)`
   - Named groups can be used as variables for replacement. In the above example the variable is: `$mynamedgroup`
 - **$d**: Enumerate every match (passing additional d's will pad the output up to a maximum of 8 digits)
 - **$dddddddd**: Example of appending the maximum padding for enumerating files

Variables can be escaped by passing `\` such as `\$0`

Specifying a `$` alone without escaping it will not have any effect on the output

## Discoverability and Convenience
A new context menu item was added to the Torrents Table to make the Multi-Rename feature more discoverable and convenient to access:  
![image](https://user-images.githubusercontent.com/3218210/216842524-450dbf63-ac08-4b27-80e7-72a326200b09.png)

By default the Multi-Rename user interface will remember the settings of the last rename operation. The following fields will be remembered: `Search`, `Use Regex`, `Match All Occurrences`, `Case Sensitive`, `Replace`, `Applies To`, `Include Files`, `Include Folders`, `Enumerate File Count`, and `Replacement Operation`. This setting can be turned off by unchecking the `Remember Multi-Rename settings` checkbox:  
![image](https://user-images.githubusercontent.com/3218210/219829091-fb2c8676-4267-45fb-94d9-48383e94b8e7.png)

ArcticGems commented:
I dont know what the developers think of this, but I think this is bloat that add unnecessary maintenance for the developers.

loligans commented:
> I dont know what the developers think of this, but I think this is bloat that add unnecessary maintenance for the developers.

I understand your concern, however I could use this feature and I'm sure many others could too. In addition, I planned on maintaining this feature.

NotTsunami commented:
> I dont know what the developers think of this, but I think this is bloat that add unnecessary maintenance for the developers.

Can you define bloat for me? You voiced the same concern in the past for a larger featureset targeting the main GUI, but I truly wonder where the line is drawn for a new feature to be considered bloat or not. This is a significantly smaller patch that I think is worth looking into.

Realistically, this will take some time to be reviewed. 5.0.0 is probably a better milestone for this PR.

thalieht commented:
Select folder + file in it
include files + folders + repalce all
rename a common part
Rename failed: file or folder already exists 'renamed_file_name' (folder succeeded)
Do note that when i was testing the above, i somehow managed to make it create a new folder with the new name and the selected file was moved in it. Unfortunately i wanted to finish with the above and i thought i'd reproduce it later with more details but now i have no idea how i did it and what settings were related.

An enhancement would be to create the window width with options on the left + longest filename x2 (so there is space for original and renamed columns).

I didn't test regex and i went in blind (didn't read OP description). I had no idea what "Enumerate Files" does and i don't see how users will figure it out without some explanation.

>Include Files: When checked, the search will apply to files regardless of any files that are checked
Include Folders: When checked, the search will apply to folders regardless of any folders are checked

Maybe i misunderstood what "the search will apply" means but i can only rename checked items.

loligans commented:
>  i somehow managed to make it create a new folder with the new name and the selected file was moved in it.

This is something I intentionally left in because the single file rename actually allows you to specify a folder path and the file will be moved into the folder path. i.e.
```
MyRootFolder
  - ChildFolder1/
  - TestFile
```
`TestFile` can be renamed to be under a subfolder by specifying `TestFolder/TestFile` for the replacement
```

MyRootFolder
  - ChildFolder1/
  - TestFolder/
    - TestFile
```
That said, I could intentionally block this behavior if it isn't what we'd like to see.  

> An enhancement would be to create the window width with options on the left + longest filename x2 (so there is space for original and renamed columns).

I'm not opposed to this, however the window is already resizable, so I'd consider this as something that would be a good candidate for a next version

> I didn't test regex and i went in blind (didn't read OP description). I had no idea what "Enumerate Files" does and i don't see how users will figure it out without some explanation.

This is a really good point and if there is a hint icon `(?)` that users can hover over with a brief description I think that might help. I tried to make the interface as intuitive as possible but multi-file renaming is a little complicated

> Maybe i misunderstood what "the search will apply" means but i can only rename checked items.

We can change the verbiage to be more concise. Correct you can only rename `checked` files. However, with `Include Files` checked and `Include Folders` unchecked, folders that have been checked will be ignored. If you want to rename folders you would need to check `Include Folders`

loligans commented:
Updated https://github.com/qbittorrent/qBittorrent/pull/18287#issue-1510232540 to include:
> Right clicking a folder will allow you to toggle all items underneath that folder. `Check` vs. `Unchecked` is determined by the row you right clicked. i.e. if the row you right clicked is `unchecked` then it will be marked `checked` and all child items will be marked `checked` as well.

Roardom commented:
> That said, I could intentionally block this behavior if it isn't what we'd like to see. I left it in because the existing single file rename allows that behavior.

Please do not block this behavior. I use it frequently when cross seeding and know others who do as well. Also linking #16906 #16834 #17049

Mika3578 commented:
Is there a way to use with a Filebot script to auto rename files in desired format ?

thalieht commented:
>TestFile can be renamed to be under a subfolder by specifying TestFolder/TestFile for the replacement

Of course that's fine by me. The thing is, there was no slash involved when i did it. In fact it happened again now that i was re-testing the PR and although this time i payed attention i can't reproduce it again. What i did:
1. Select root folder + a file in it.
2. Open rename dialog
3. Fill the fields for search and replace
4. Tick `Include folders`
5. Select `Replace all` and click it

Ranamed `milo` and `file1.rar` (root folder and file) to for example replace "il" with "on" (there was no `folder1` last time, every file was in the root folder)
```
milo <----root folder
    folder1
        file1.rar
    file2
    file3
```

It became:
```
milo
    folder1
        fone1.rar
mono <---  (milo -> mono)
    file2
    file3
```

You didn't comment on the problem i described with bullet points at the beginning of https://github.com/qbittorrent/qBittorrent/pull/18287#issuecomment-1364697445 which still happen by the way. It seems to be related with the unwanted folder creation described above because when that happened there was no error,

Another problem that i just noticed is that the `Renamed` column header does not appear when horizontal scrolling and it's out of view. It appears when window is resized so there is space for it.

One last thing: any existing error from previous rename isn't cleared after a successful rename (in the same dialog).

loligans commented:
>  In fact it happened again now that i was re-testing the PR and although this time i payed attention i can't reproduce it again.

Interesting, would you mind uploading a video/gif of the behavior you're experiencing? One potential thing to test for reproducibility is refreshing the page before each test.

> Ranamed milo and file1.rar (root folder and file) to for example replace "il" with "on" (there was no folder1 last time, every file was in the root folder)

This suggests that order of operations matter's here. i.e. `milo` was renamed to `mono` however, `file1.rar` still contained the old `path` and as a result the rename operation put it under a new `milo` folder. I'll see how I can fix this.

> You didn't comment on the problem i described with bullet points at the beginning of https://github.com/qbittorrent/qBittorrent/pull/18287#issuecomment-1364697445 which still happen by the way. It seems to be related with the unwanted folder creation described above because when that happened there was no error,

I think you're correct about it being related. Essentially the rename all operation changes files one by one and if it encounters an error it will fire an event and continue renaming. 

> Another problem that i just noticed is that the Renamed column header does not appear when horizontal scrolling and it's out of view. It appears when window is resized so there is space for it.

The columns remember their original position by default, I think this can be changed. such that both share an equal amount of horizontal space.

> One last thing: any existing error from previous rename isn't cleared after a successful rename (in the same dialog).

good catch haha I'll fix this.

thalieht commented:
>Interesting, would you mind uploading a video/gif of the behavior you're experiencing?

Unfortunately i can't reproduce it again.

>>Another problem that i just noticed is that the Renamed column header does not appear when horizontal scrolling and it's out of view. It appears when window is resized so there is space for it.
>
>The columns remember their original position by default, I think this can be changed. such that both share an equal amount of horizontal space.

![header](https://user-images.githubusercontent.com/6451685/211197511-cef0992a-fc88-43d4-911a-3307099fe4ef.gif)

loligans commented:
@thalieht Okay I've pushed up some changes to fix the issues you were running into. I'll cover the changes in this comment.

1. Added a `remove` function to the `LocalPreferencesClass` so that I could remove localStorage keys.
2. Prevented the BulkRename table from remembering the column widths by removing the key in localStorage that stores the columns previous width. The reason is because the window width is not remembered by default.
3. Fixed `Rename All` by iterating in reverse, this is because the `matchedFiles` are in DFS order which should prevent unwanted file/folder creation. _Note: Single File Renames still iterate linearly but since the table gets refreshed after every rename we don't run into the problem of unwanted files/folders._
4. Fixed the header scrolling by binding the table body horizontal scroll location to the table header horizontal scroll location. 

Let me know if I missed anything or if my assumptions are incorrect and of course if you or anyone else run into a bug with the updated code.

thalieht reviewed (dismissed):
LGTM (as a tester)

loligans commented:
Pushed up changes to fix the WebUI linter errors generated from the linter workflow

loligans commented:
i'm Not sure who from the _Web developers_ group can review this PR but i believe it is ready for a review. @glassez or @thalieht maybe you know? i'm mostly concerned about what are the next steps i need to take in order to get this PR ready to be merged into `master`.

glassez commented:
@loligans 
You don't need to rebase and force push the PR every time `master` is updated unless there are conflicts. Otherwise, it seems that you have made some new changes that need to be reviewed again.

loligans commented:
I've been using this branch on my server for a bit and added some quality of life improvements. Here are the latest updates:

- Added `Rename Files...` to the Torrents Table to make the Multi-Rename feature more discoverable and convenient to access.
![image](https://user-images.githubusercontent.com/3218210/216842524-450dbf63-ac08-4b27-80e7-72a326200b09.png)

- Added ability to remember preferences between renaming operations.
  - The following fields are remembered: `Search`, `Use Regex`, `Match All Occurrences`, `Case Sensitive`, `Replace`, `Applies To`, `Include Files`, `Include Folders`, `Enumerate File Count`, and `Replacement Operation`.
- Added option to the preferences window for toggling remembering Multi-Rename  preferences between renaming operations (Defaults to `true` i.e. to remembering the preferences by default). This preference gets saved to `LocalPreferences`
![image](https://user-images.githubusercontent.com/3218210/216842598-3ca75897-fa06-472b-a33a-b89d83f35dd0.png)

- Fixed a bug when the Replacement Operation was Single File and File Enumeration Count was enabled, every Rename operation would reset the file count for subsequent files. This causes file name conflicts, so now the File Enumeration Count increments by 1 after every single file rename operation.

glassez commented:
> * Added `Rename Files...` to the Torrents Table to make the Multi-Rename feature more discoverable and convenient to access.

It's questionable...
How can someone want to rename files without first looking at the appropriate place where they are listed? So accessing this function from the file list is enough, isn't it?

glassez commented:
New changes require new review/test. @qbittorrent/bug-handlers?

loligans commented:
> It's questionable... How can someone want to rename files without first looking at the appropriate place where they are listed? So accessing this function from the file list is enough, isn't it?

What I've discovered is that with some torrents you can infer the file layout without looking at the files. However regardless of being able to infer the file layout, when you select the `Rename Files...` context menu item, it will open the Multi-Rename window with all the files listed but none of them selected.

The feature is more of a convenience for people who don't want to have to: go to the torrent files table, select 2 or more files, right-click, and select Rename.

thalieht commented on src/webui/www/private/views/preferences.html:
```diff
@@ -122,6 +122,12 @@
+            <tr>
+                <td>
+                    <input type="checkbox" id="multirename_rememberprefs_checkbox" onchange="LocalPreferences.set('multirename_rememberPreferences', this.checked);" />
+                    <label for="multirename_rememberprefs_checkbox">QBT_TR(Remember Multi-Rename Preferences)QBT_TR[CONTEXT=OptionsDialog]</label>
```
About capitalization, not sure if "Multi-Rename" is special but "Preferences" is an ordinary word and other settings use lower case
```suggestion
                    <label for="multirename_rememberprefs_checkbox">QBT_TR(Remember Multi-Rename preferences)QBT_TR[CONTEXT=OptionsDialog]</label>
```

/Off topic for this line
You should disable/remove "Rename files..." entry in torrent list's context menu when torrent doesn't have metadata and when multiple torrents are selected.

Other than that LGTM.

loligans replied:
Fixed the capitalization and added a check that will hide `Set Location...`, `Rename...`, and `Rename Files...` when multiple torrents are selected

glassez replied:
Why `Set Location`?

loligans replied:
When multiple torrents were selected, the `Select Location...` menu item had no effect

glassez replied:
Seems you took it incorrectly. It just extracts current save path from the first selected torrent. But it send all the selected torrents hashes:
https://github.com/loligans/qBittorrent/blob/0939417f1b225758f24df49867a7ca657fc0156c/src/webui/www/private/scripts/mocha-init.js#L492

glassez replied:
@loligans 
General comment is making unrelated changes is the best way to slow down your PR.

loligans replied:
my mistake, i'll undo that change.

loligans replied:
The change has been reverted, now the only menu item's that get hidden are `Rename...` and `Rename Files...` when multiple torrents are selected

glassez replied:
> when multiple torrents are selected

What about the case when single torrent is selected but it has no metadata?

loligans replied:
i can make that change too, although i'm not sure where i can check that a torrent has metadata. 

i'm not familiar with torrent metadata and so i could use a little help with knowing where to look on the webui

glassez replied:
> i can make that change too, although i'm not sure where i can check that a torrent has metadata.

It is not explicitly represented via WebAPI. However, since a torrent without metadata doesn't know anything about its payload, you can estimate some kind of indirect field, such as `total_size` which is `-1` in case when there is no metadata available.
Or just use the way suggested by @thalieht in https://github.com/qbittorrent/qBittorrent/pull/18287#discussion_r1103803194.

loligans replied:
thanks! i'll look at both options and add the functionality.

glassez commented:
> The feature is more of a convenience for people who don't want to have to: go to the torrent files table, select 2 or more files, right-click, and select Rename.

I understand your motivation. But the flip side of the coin is that adding more and more menu items reduces its convenience, right? However, if no one objects, I don't insist.

loligans commented:
> I understand your motivation. But the flip side of the coin is that adding more and more menu items reduces its convenience, right? However, if no one objects, I don't insist.

I think that's a fair and objective assessment. I agree with you that if more people weigh in saying that they don't want the functionality then I would have no problem removing it now or in a future PR.

thalieht commented on src/webui/www/private/views/transferlist.html:
```diff
@@ -106,6 +109,14 @@
                 y: 2
+            },
+            onShow: function() {
+                // Hides rename, and renameFiles menu items when multiple torrents are selected
```
I don't know if this is the right place to do this but other items in the context menu are hidden in the function below
(this line points on how to check for metadata)
`if (data['state'] == 'metaDL' || data['state'] == 'forcedMetaDL')`
https://github.com/qbittorrent/qBittorrent/blob/915121a0dd2cedad2912eddea194c53bf4b5500c/src/webui/www/private/scripts/contextmenu.js#L335

loligans replied:
thanks! i'll get this added

loligans replied:
i added a change to only allow renaming when 1 item is selected and also when metadata is `not` downloaded

https://github.com/loligans/qBittorrent/blob/f563267f0bac867d008f06fb81aabf67c19c0b23/src/webui/www/private/scripts/contextmenu.js#L365-L372

loligans replied:
my mistake, i didn't know of a good way to create a torrent that hadn't downloaded metadata. i figured out how and got it squared away. on a side note, i only hide the `renameFiles` menu item now because `rename` can be done on a torrent that is still downloading metadata which i was unaware of. i don't plan on touching anything else for the rest of this PR

glassez commented:
> This setting can be turned off in the `Saving Management` preferences menu:
> ![image](https://user-images.githubusercontent.com/3218210/216842598-3ca75897-fa06-472b-a33a-b89d83f35dd0.png)

"Saving management" is about torrent content and related things.
IMO, the best place for such checkbox is the "rename dialog" itself.

Traineratwot commented:
@loligans I need this feature, I have a TV series with more than 1000 episodes and there is a jellyfin program that better recognizes files named in a certain way, and I would like to be able to rename from and still remain a torrent distributor. can I repeat your preview? it works? 
p.s  
I think everyone here underestimates this functionality

loligans commented:
> "Saving management" is about torrent content and related things.
> 
> IMO, the best place for such checkbox is the "rename dialog" itself.

sounds good, i'm working on a change to move it into the rename dialog.

loligans commented:
@glassez the latest change moves the preference to the renaming window:
![image](https://user-images.githubusercontent.com/3218210/219829091-fb2c8676-4267-45fb-94d9-48383e94b8e7.png)

loligans commented:
> @loligans I need this feature, I have a TV series with more than 1000 episodes and there is a jellyfin program that better recognizes files named in a certain way, and I would like to be able to rename from and still remain a torrent distributor. can I repeat your preview? it works? p.s I think everyone here underestimates this functionality

the functionality outlined in this pr will allow you to rename multiple files of a torrent as a bulk operation and still be able to remain a torrent distributor. the preview i have been using confirms this behavior

glassez commented:
@loligans 
Thank you!

Note that some flaws may remain hidden until it goes in the release so that many users will have access to it. I hope you'll still be here at that moment to fix them.

loligans commented:
yup, i plan to support this feature going forward!

varad-kadam commented:
Hello,
I'm using `qbittorrent-nox`, and on `v4.5.4 Web UI (64-bit)`.
From what I see, this feature being tested and set to be introduced in `v5.0.0`, however I really need it now.

Could someone please detail on how to build / get a custom version of `qbittorrent-nox` working that includes this patch?

PriitUring commented:
@varad-kadam

This is got merged into Master branch ~5 months ago, has Milestones tag 4.6.0, so this should be included in many already publicly released qBit v4.6.0 test builds and I'm unable to see v5 mentioned anywhere here, so it's weird where did you find something else being written.

If you need help building qBit nox-static, then better to ask in nox discussions page:
https://github.com/userdocs/qbittorrent-nox-static/discussions

JDENredden commented:
Can this be automated to act on every torrent when added? Can I call it from the API?

loligans commented:
> Can this be automated to act on every torrent when added? Can I call it from the API?

No this currently is implemented on the frontend using JavaScript. It cannot be invoked by any API on the qBittorrent backend.

boydaihungst commented:
> Being a complete novice, how do I add this great functionality?

Use version >= 4.6.0. Right-click torrent > rename files. Or select >= 2 files in the Content tab.
Tips: this can also be used to move files to sub-folders: `folder1/folder2/folder3/$0`

Anrock commented:
Would be nice to have help texts (the ones from description are good) as tooltips for input fields and checkboxes. I've found this PR via google when trying to understand how to use this feature.
